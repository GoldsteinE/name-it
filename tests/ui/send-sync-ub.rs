use name_it::name_it;

use futures::prelude::*;
use std::{cell::Cell, rc::Rc, sync::atomic, thread};
type X = Result<&'static [u8], &'static [()]>;

#[name_it(Foo)]
async fn foo(x: Rc<Cell<X>>) -> Rc<Cell<X>> {
    x
}

fn main() {
    let cell = Rc::new(Cell::new(Ok(&[][..])));
    let fut = foo(cell.clone());

    // Foo implements Send, even though it contains an Rc<Cell<...>>
    thread::spawn(|| {
        // get the Rc<Cell<...>> back out of the future
        let cell = fut.now_or_never().unwrap();

        // create a data race
        loop {
            atomic::fence(atomic::Ordering::SeqCst); // make this exploit work with --release, too

            cell.set(Ok(&[][..])); // zero-element `u8` array

            atomic::fence(atomic::Ordering::SeqCst); // make this exploit work with --release, too

            cell.set(Err(&[(); 1000000][..])); // very long `()` array
        }
    });

    // create a data race
    loop {
        atomic::fence(atomic::Ordering::SeqCst); // make this exploit work with --release, too

        // check for `u8` slice with len of the `()` slice (and pointer to either one)
        if let Ok(slice @ [_, ..]) = cell.get() {
            for i in (0..1000000).step_by(10000) {
                dbg!(slice[i]); // create a seg-fault
            }
            return;
        }
    }
}
