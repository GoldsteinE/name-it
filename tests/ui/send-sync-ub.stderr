error[E0277]: `*const ()` cannot be sent between threads safely
   --> tests/ui/send-sync-ub.rs:17:5
    |
17  |     thread::spawn(|| {
    |     ^^^^^^^^^^^^^ -- within this `[closure@$DIR/tests/ui/send-sync-ub.rs:17:19: 17:21]`
    |     |
    |     `*const ()` cannot be sent between threads safely
    |
    = help: within `[closure@$DIR/tests/ui/send-sync-ub.rs:17:19: 17:21]`, the trait `std::marker::Send` is not implemented for `*const ()`
    = note: required because it appears within the type `PhantomData<*const ()>`
    = note: required because it appears within the type `Markers<false, false, false>`
note: required because it appears within the type `_foo_impl::Foo<'_>`
   --> tests/ui/send-sync-ub.rs:7:11
    |
7   | #[name_it(Foo)]
    |           ^^^
    = note: required because it appears within the type `Named<_foo_impl::Foo<'_>>`
note: required because it's used within this closure
   --> tests/ui/send-sync-ub.rs:17:19
    |
17  |     thread::spawn(|| {
    |                   ^^
note: required by a bound in `std::thread::spawn`
   --> $RUST/std/src/thread/mod.rs
    |
    |     F: Send + 'static,
    |        ^^^^ required by this bound in `std::thread::spawn`
